src/controller/couponController.ts

```diff
@@ -0,0 +1,143 @@
+            while (await couponRepository.findOne({ where: { code } })) {
+                code = crypto.randomBytes(4).toString('hex');
+            }
+            const coupon = new Coupon({ 
+                code,
+                description, 
+                percentage, 
+                expirationDate, 
+                applicableProducts: products
+             });
+            const newCoupon = await couponRepository.save(coupon);
+            res.status(201).json({
+                message: 'Coupon created successfully',
+                data: newCoupon
+            });
+        })
+    ]
+
+    // GET /coupons/:id
+    public getCouponById = errorHandler(async(req: Request, res: Response) => {
+        const { id } = req.params;
+        const coupon = await couponRepository.findOne({ 
+            where: { id: Number(id) }, 
+            relations: ['applicableProducts']
+        });
+        if (!coupon) {
+            return res.status(404).json({ error: 'Coupon not found' });
+        } else {
+            return res.status(200).json(coupon);
+        }
+    })
+
+    // PUT /coupons/:id
+    public updateCoupon = [
+        ...createCouponRules,
+        errorHandler(async(req: Request, res: Response) => {
+            const errors = validationResult(req);
+            if (!errors.isEmpty()) {
+              return res.status(400).json({ errors: errors.array() });
+            }
+            const { id } = req.params;
+            const { description, percentage, expirationDate, applicableProducts  } = req.body as couponRequestBody;
+            const coupon = await couponRepository.findOne({
+                where: { id: Number(id) },
+            });
+            if (!coupon) {
+                return res.status(404).json({ error: 'Coupon not found' });
+            }
+            coupon.description = description;
+            coupon.percentage = percentage;
+            coupon.expirationDate = expirationDate;
+            const products: Product[] = []
+            for (const productId of applicableProducts) {
+                const product = await productRepository.findOne({
+                    where: { id: productId },
+                    relations: ['vendor']
+                });
+                if (!product) {
+                    return res.status(400).json({ error: `Product with id ${productId} not found` });
+                }
+                if (product.vendor.id !== (req.user as UserModel).id) {
+                    return res.status(403).json({ error: 'You can only create coupons for your own products' });
+                }
+                products.push(product);
+            }
+            coupon.applicableProducts = products;
+            const updatedCoupon = await couponRepository.save(coupon)
+            return res.status(200).json(updatedCoupon);
+        })
+    ]
+
+    // DELETE /coupons/:id
+    public deleteCoupon = errorHandler(async(req: Request, res: Response) => {
```

Comment:
Shouldn't you check to see if the coupon to be deleted actually belongs to the currently logged in user?

that way, a seller can only delete coupons that belong to them

Reply:
It's done